### DISCUSSION TITLE
Help Converting sklearn Time Series Feature Engineering with Repeating Basis Functions (RBF for sklego)

### DISCUSSION THREAD

**Discussion by MarkusSagen** [alan-turing-institute/sktime#7234] [Q&A]
I have the following sklearn and sklego simple linear regression forecasting method that I'd be interested in porting over to sktime. In sklearn, this existing code works as expected, but not in sktime

### Pre-requisites
```
pip install pandas scikit-learn scikit-lego sktime 
```

### Data

Found at https://github.com/MarkusSagen/sktime_example_data/blob/main/data.csv 
Preview of the data

```md
|    | ds         |       y |
|---:|:-----------|--------:|
|  0 | 2021-10-30 | 240.764 |
|  1 | 2021-10-31 | 251.31  |
|  2 | 2021-11-01 | 252.477 |
|  3 | 2021-11-02 | 239.092 |
|  4 | 2021-11-03 | 236.408 |
```

### Code

The following code works fine in sklearn, but not when converting to sktime

```python
import pandas as pd
from sklego.preprocessing import RepeatingBasisFunction
from sklearn.linear_model import LinearRegression
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split

class TimeSeriesFeatureEngineer:
    def __init__(self, date_column: str = 'ds'):
        self.date_column = date_column
    
    def transform(self, df: pd.DataFrame) -> pd.DataFrame:
        df = df.copy()
        df['day'] = (df[self.date_column] - df[self.date_column].min()).dt.days
        df['day_of_week'] = df[self.date_column].dt.dayofweek
        df['day_of_month'] = df[self.date_column].dt.day
        df['day_of_year'] = df[self.date_column].dt.dayofyear
        df['month'] = df[self.date_column].dt.month
        df['week'] = df[self.date_column].dt.isocalendar().week
        return df

data = pd.read_csv("data.csv", parse_dates="ds")

feature_engineer = TimeSeriesFeatureEngineer()
X = feature_engineer.transform(data)
y = X['y']
X = X.drop(['ds', 'y'], axis=1)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, shuffle=False)

# Create ColumnTransformer with RBFs for various features
rbf_transformer = ColumnTransformer([
    ('day_of_week_rbf', RepeatingBasisFunction(n_periods=7, input_range=(1, 7)), ['day_of_week']),
    ('day_of_month_rbf', RepeatingBasisFunction(n_periods=31, input_range=(1, 31)), ['day_of_month']),
    ('month_rbf', RepeatingBasisFunction(n_periods=12, input_range=(1, 12)), ['month']),
    ('day_passthrough', 'passthrough', ['day'])
])

pipeline = Pipeline([
    ('rbf_transformer', rbf_transformer),
    ('linear_regression', LinearRegression())
])

pipeline.fit(X_train, y_train)

# Make predictions on both train and test data
y_pred_train = pipeline.predict(X_train)
y_pred_test = pipeline.predict(X_test)
```

**Comment by MarkusSagen** [comment]
Similar issue was raised on Github, but I would not replicate the example
https://github.com/sktime/sktime/discussions/4375

**Comment by MarkusSagen** [comment]
I could get it sort-off to work by wrapping the values into dataframes

```python
# Create ColumnTransformer with RBFs for various features
rbf_transformer = ColumnTransformer([
    ('day_of_week_rbf', RepeatingBasisFunction(n_periods=7, input_range=(1, 7)), ['day_of_week']),
    ('day_of_month_rbf', RepeatingBasisFunction(n_periods=31, input_range=(1, 31)), ['day_of_month']),
    ('month_rbf', RepeatingBasisFunction(n_periods=12, input_range=(1, 12)), ['month']),
    ('week_rbf', RepeatingBasisFunction(n_periods=52, input_range=(1, 52)), ['week']),
    ('day', RepeatingBasisFunction(n_periods==365, input_range=(1, 366)), ["day"]),
    ('day_passthrough', 'passthrough', ['day'])
])

Z_train = rbf_transformer.fit_transform(X_train)
Z_test = rbf_transformer.transform(X_test)

f = make_reduction(LinearRegression(), strategy="recursive")
f.fit(y=pd.DataFrame(y_train), X=pd.DataFrame(Z_train), fh=[1, 2, 3])
y_pred_train = f.predict(fh=list(range(1, 120)), X=pd.DataFrame(Z_train))
y_pred_test = f.predict(fh=list(range(1, 120)), X=pd.DataFrame(Z_test))
```

But the values are WAY off then

**Comment by fkiraly** [comment]
I think I know what you want to do, please confirm if I understand correctly - you want to use linear regression to forecast the value from date-time features at the same future time using a linear model, right? This is a useful baseline model.

The way you would do it in `sktime` uses less code than the `sklego` example:

```python
from sktime.transformations.series.date import DateTimeFeatures
from sktime.forecasting.compose import YfromX
from sklearn.linear_model import LinearRegression

basis_f = DateTimeFeatures(feature_scope="efficient")
fcst = basis_f ** YfromX(LinearRegression())
```

here we use that the `"efficient"` feature scope seems to agree mostly with your selection of dummies. To remove some, you can use subsetting notation or the args of `DateTimeFeatures`.

The `**` applies the `basis_f` to the exogeneous data, i.e., it generates feature sthere.

To include lags manually, you can do something like

```python
from sktime.transformations.series.date import DateTimeFeatures
from sktime.transformations.series.lag import Lag
from sktime.forecasting.compose import YfromX
from sklearn.linear_model import LinearRegression

basis_f = DateTimeFeatures(feature_scope="efficient")
lags = Lag([1, 2, 7, 30])
fcst = (basis_f + lags) ** YfromX(LinearRegression())
```

You can also do `make_reduction` and chain the `basis_f` with it, instead.

**Reply by fkiraly** [reply]
PS: The data needs to be in `sktime` format, so starting from your csv you need to do `set_index("ds")` first. Might be nice if you could include some data generation code with a dummy dataset.

**Comment by fkiraly** [comment]
@MarkusSagen, this was implemented by @phoeenniixx in https://github.com/sktime/sktime/pull/7261, and is now out in the 0.33.2 release!

**Reply by MarkusSagen** [reply]
I'm curious how this would work as a complete example

So far, I've not been able to make it work with a similar pipeline setup as scikit-lego described

**Reply by fkiraly** [reply]
```python
from sktime.datasets import load_airline
from sktime.transformations.series.basisfunction import RBFTransformer
from sktime.forecasting.compose import YfromX
from sklearn.linear_model import LinearRegression

basis_f = RBFTransformer()
fcst = basis_f ** YfromX(LinearRegression())

y = load_airline()
fcst.fit(y, fh=[1, 2, 3])
fcst.predict()
```

This will result in bad forecasts, but you need to set centers and gamma.

FYI @phoeenniixx, I think it would be important to say what the unit of the `centers` and `gamma` is in the datetime case, especially if you pass integers.

**Reply by phoeenniixx** [reply]
yes, to get the "good" forecasts, you must set centers and gamma, they should not be given any random value, rather you should keep some things in mind:
- the `centers ` should cover both historical data and forecast horizon.
- Choose `gamma` based on your desired RBF width:

     -Too large: RBFs become very narrow, leading to sparse features
     -Too small: RBFs become too wide, losing temporal precision
     -Rule of thumb: gamma = 1/(expected_time_scale^2)

centers: When working with time series data:

For datetime indices: Centers are internally converted to numeric values using `pd.to_numeric()`. So you should either:

   a) Provide center points as datetime values which will be converted automatically, or
   b) Provide numeric values that match the scale of your converted datetime index

For numpy arrays: Centers are compared against simple positional indices (0, 1, 2...), so provide centers that match this scale


gamma: For datetime indices, gamma should be scaled appropriately for the time scale of your data:

For daily data: gamma = 1/(n_days^2) where n_days is your expected time scale
For monthly data: gamma = 1/(n_months^2)
For yearly data: gamma = 1/(n_years^2)
